### ISSUE TITLE
Expect that a  mocked function is never called

### ISSUE THREAD

**Issue by jgashe** [plataformatec/mox#32]
Mox requires that, if a mock is used, we set an expectation for it. But `expect` requires that the number of invocations be greater than 0. Other mock libraries (like [mockito](http://static.javadoc.io/org.mockito/mockito-core/2.15.0/org/mockito/Mockito.html#exact_verification)) allow for verifying that a mocked method is never called. This would probably be useful for Mox.

Thoughts?

**Comment by josevalim** [comment]
@jgashe if you try to call something that has no expectation it will just raise today. And your test should fail accordingly.

**Comment by jgashe** [comment]
I'd like to have a test pass if something is *never* called. The function never being called is my expected behavior.

Imagine code like this:
```
def conditional_call(my_condition) do
  if (my_condition) do
    MyMock.call()
  else
    do_nothing
  end
end
```

If I want to test both branches of this, I'd call `conditional_call` with both true and false. If I'm writing a test for calling `conditional_call` with `false`, I'd want to expect that MyMock.call is never called, and have the test fail if MyMock.call is called. 

Mox doesn't let me do this because it raises if the Mock doesn't have any expectation, and doesn't have a pattern match of `expect` where n = 0.

**Comment by josevalim** [comment]
> Mox doesn't let me do this because it raises if the Mock doesn't have any expectation

What do you mean? If you don't set any expectation that is fine by Mox.

**Comment by jgashe** [comment]
Oh yeah you're totally right.

Closed.

**Comment by josevalim** [comment]
@NikitaAvvakumov if the process is not linked to the test process in any way, then the failures won't cascade, and your test won't fail. this is a general property of Elixir, not necessarily related to mox.

**Comment by NikitaAvvakumov** [comment]
@josevalim But in this particular case, could test failures be enforced by adding to Mox either `expect(m, f, 0, _fn)` or a new function like `refute` or `expect_not`, like the OP was asking? If such existed, it could be combined with allowances or global mode and `:verify!`/`verify_on_exit!` to cause tests to fail and, arguably, make their intentions more explicit.

**Comment by josevalim** [comment]
@NikitaAvvakumov oh, I see what you mean. That sounds good to me. Can you please send a PR?

**Comment by NikitaAvvakumov** [comment]
@josevalim I'll gladly give it a shot. Which option sounds better to you - `expect` accepting `n = 0` or `refute` or similarly named new function?

**Comment by josevalim** [comment]
`expect` with 0.

**Comment by steffkes** [comment]
@josevalim just reading through older issues. this seems to be addressed in #35 which is already merged, right?

**Comment by josevalim** [comment]
correct! thanks!
